src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -25,59 +25,51 @@ public InstalledAndTransitivePackageFeed(IEnumerable<PackageCollectionItem> inst
         /// <inheritdoc cref="IPackageFeed.ContinueSearchAsync(ContinuationToken, CancellationToken)" />
         public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchAsync(ContinuationToken continuationToken, CancellationToken cancellationToken)
         {
-            var searchToken = continuationToken as FeedSearchContinuationToken;
-            if (searchToken == null)
-            {
-                throw new InvalidOperationException(Strings.Exception_InvalidContinuationToken);
-            }
+            var searchToken = ThrowIfNotFeedSearchContinuationToken(continuationToken);
+            cancellationToken.ThrowIfCancellationRequested();
 
             // Remove transitive packages from project references
             IEnumerable<PackageCollectionItem> pkgsWithOrigins = _transitivePackages
                 .Where(t => t.PackageReferences.Any(x => x is ITransitivePackageReferenceContextInfo y && y.TransitiveOrigins.Any()));
 
-            IEnumerable<PackageCollectionItem> pkgs = _installedPackages.Concat(pkgsWithOrigins);
-            PackageCollectionItem[] allPkgs = PerformLookup(pkgs, searchToken);
+            IPackageSearchMetadata[] intalledItems = await DoSearchAsync(_installedPackages, searchToken, cancellationToken);
+            IPackageSearchMetadata[] transitiveItems = await DoSearchAsync(pkgsWithOrigins, searchToken, cancellationToken);
```

Comment:
DoSearchAsync handles two responsibilities: Performing the lookup and for the results, get the package metadata for each item. This means we have to pass in parameters for both of these operations to this method.

What if instead, we called PerformLookup and passed that into a new method that does what the second part of DoSearchAsync does so and we can call it something like "GetMetadataForPackagesAsync" and we can remove the DoSearchAsync method. Then we could call it like this:

```suggestion
            IPackageSearchMetadata[] transitiveItems = await GetMetadataForPackagesAsync(PerformLookup(pkgsWithOrigins, searchToken), cancellationToken);
```

That way, our new method has only one responsibility and its name can clearly reflect its responsibility, and it makes it easier to test because one of its parameters is an array of items T[].

Reply:
Splitting this might be fine, but the proposal for `PerformLookup` I believe would mean we're synchronously string-searching and sorting a list of installed packages with LINQ.  Is that acceptable for the worst-case # installed packages?

Reply:
PerformLookup already exists as a synchronous method so whether it's called from within another async method which is awaited here or from within this asynchronous method directly, the result should be the same considering we have the await on DoSearch.

Reply:
Created two new methods.